src/mds/MDSRank.cc

```diff
@@ -1640,18 +1640,12 @@ void MDSRank::replay_start()
   calc_recovery_set();
 
   // Check if we need to wait for a newer OSD map before starting
-  Context *fin = new C_IO_Wrapper(this, new C_MDS_BootStart(this, MDS_BOOT_INITIAL));
-  bool const ready = objecter->wait_for_map(
-      mdsmap->get_last_failure_osd_epoch(),
-      fin);
-
-  if (ready) {
-    delete fin;
-    boot_start();
-  } else {
-    dout(1) << " waiting for osdmap " << mdsmap->get_last_failure_osd_epoch()
-	    << " (which blacklists prior instance)" << dendl;
-  }
+  auto fin = new C_IO_Wrapper(this, new C_MDS_BootStart(this, MDS_BOOT_INITIAL));
```

Comment:
...warms the heart, it does. Though if we're just using std::ref() why the ptr?

Reply:
Same reason. I was trying to minimize the amount of time I would need to spend rewriting the MDS. In retrospect it might be better to use an explicit lambda.

Reply:
There. Replaced these ugly std::ref hacks by wrapping the Context*s up in lambda that hold a unique_ptr.

And replaced the uses of use_future with ceph::waiter.

Not that there's anything wrong with futures, but each future inherently allocates a chunk of memory that future and promise point to so they can both be moved around and returned from things. You can't pass waiters around, so no allocation needed.

Aaaand. I think that is all of the Ugliness that you and others have drawn my attention to. Thanks!

(If you review more, start a new review, hopefully it'll show up at the bottom of the page rather than being collapsed and hidden away.)

Reply:
+1 great work! :-)